core/src/env/engine/off_chain/chain_extension.rs

```diff
@@ -0,0 +1,112 @@
+use crate::env::Result;
+use std::collections::HashMap;
+
+type FuncId = u32;
+
+/// Types implementing this trait can be used as chain extensions.
+///
+/// This trait is only useful for testing contract via the off-chain environment.
+pub trait ChainExtension {
+    /// The expected input type.
+    ///
+    /// # Note
+    ///
+    /// This can be a tuple to expect multiple input types.
+    type Input: scale::Codec;
+    /// The expected output type.
+    type Output: scale::Codec;
+
+    /// The static function ID of the chain extension.
+    ///
+    /// # Note
+    ///
+    /// This is expected to return a constant value.
+    fn func_id(&self) -> u32;
+
+    /// Calls the chain extension with the given input.
+    fn call(&mut self, input: &Self::Input) -> Result<Self::Output>;
+}
+
+/// A raw chain extension function.
+///
+/// This is mostly a wrapper closure around the real chain extension function
+/// that handles marshalling of types between their encoded and decoded
+/// representations.
+type ChainExtensionFn = Box<dyn FnMut(Vec<u8>) -> Result<Vec<u8>>>;
+
+/// Runtime call handler.
+///
+/// More generically a mapping from bytes to bytes.
+pub struct ChainExtensionHandler {
+    /// The currently registered runtime call handler.
+    registered: HashMap<FuncId, ChainExtensionFn>,
+}
+
+impl ChainExtensionHandler {
+    /// Creates a new chain extension handler.
+    ///
+    /// Initialized with an empty set of chain extensions.
+    pub fn new() -> Self {
+        Self {
+            registered: HashMap::new(),
+        }
+    }
+
+    /// Resets the chain extension handler to uninitialized state.
+    pub fn reset(&mut self) {
+        self.registered.clear()
+    }
+
+    /// Register a new chain extension.
+    pub fn register<I, O>(
+        &mut self,
+        mut extension: Box<dyn ChainExtension<Input = I, Output = O>>,
+    ) where
+        I: scale::Codec + 'static,
+        O: scale::Codec + 'static,
+    {
+        let func_id = extension.func_id();
+        self.registered.insert(
+            func_id,
+            Box::new(move |encoded_input: Vec<u8>| {
+                let decoded_input = scale::Decode::decode(&mut &encoded_input[..])?;
+                let decoded_output = extension.call(&decoded_input)?;
+                Ok(scale::Encode::encode(&decoded_output))
+            }),
+        );
+    }
+
+    /// Invokes the chain extension with the given parameters.
+    pub fn invoke<I, O>(&mut self, func_id: FuncId, input: &I) -> Result<O>
+    where
+        I: scale::Codec + 'static,
+        O: scale::Codec + 'static,
+    {
+        use std::collections::hash_map::Entry;
+        match self.registered.entry(func_id) {
+            Entry::Occupied(mut occupied) => {
+                let encoded_input = scale::Encode::encode(input);
+                let encoded_output = occupied.get_mut()(encoded_input)?;
+                scale::Decode::decode(&mut &encoded_output[..]).map_err(Into::into)
```

Comment:
Why to encode it here just to have it decoded again in the closure in order to pass it to the Trait implementer? Why not just pass it through here and remove the decoding pass in the closure?

Reply:
The problem is that the off-chain environment is a static object that cannot be generic so it needs to marshall all values as it would be done in the real Seal. I see that this is bad but unfortunately I haven't come up with a better idea.